storage/innobase/row/row0purge.cc

```diff
@@ -943,10 +954,27 @@ static trx_id_t row_purge_remove_sec_if_poss_leaf(purge_node_t *node,
 			}
 
 			if (btr_cur_optimistic_delete(btr_cur, 0, &mtr)
-                            == DB_FAIL) {
+			    == DB_FAIL) {
 				page_max_trx_id = page_get_max_trx_id(
 					btr_cur_get_page(btr_cur));
-                        }
+				trx_t* trx = purge_sys.query->trx;
+				if (trx->max_inactive_id < page_max_trx_id) {
+					trx->mutex_lock();
+					bool hot = trx_sys.find_same_or_older(
+						trx, page_max_trx_id);
+					trx->mutex_unlock();
+					if (hot) {
+						/* Since an active transaction
+						may modify the secondary index
+						but not PAGE_MAX_TRX_ID,
+						row_purge_poss_sec() must be
+						invoked again after re-latching
+						the index page. Let us return
+						an impossible ID. */
+						page_max_trx_id = ~trx_id_t{0};
+					}
+				}
+			}
```

Comment:
Come to think of it, all of this code should maybe placed in an `ATTRIBUTE_COLD ATTRIBUTE_NOINLINE static` function. I’ll check the generated code once more.

Moreover, I think that `1` should be a more convenient "impossible" return value, because then the value could be assigned from a register that holds the value of `hot`. Transaction 1 would be the following during InnoDB bootstrap:
```
#0  trx_sys_t::get_new_trx_id_no_refresh (this=0x557174a4c180 <trx_sys>) at /mariadb/10.6/storage/innobase/include/trx0sys.h:1257
#1  trx_sys_t::register_rw (this=0x557174a4c180 <trx_sys>, trx=trx@entry=0x7f0fb0e023c0) at /mariadb/10.6/storage/innobase/include/trx0sys.h:1114
#2  0x000055717316ce15 in trx_assign_rseg_low (trx=trx@entry=0x7f0fb0e023c0) at /mariadb/10.6/storage/innobase/trx/trx0trx.cc:811
#3  0x000055717316e27f in trx_start_low (trx=trx@entry=0x7f0fb0e023c0, read_write=read_write@entry=0x1) at /mariadb/10.6/storage/innobase/trx/trx0trx.cc:953
#4  0x000055717316e988 in trx_start_internal_low (trx=trx@entry=0x7f0fb0e023c0, read_write=read_write@entry=0x1) at /mariadb/10.6/storage/innobase/trx/trx0trx.cc:2263
#5  0x000055717316e9c5 in trx_start_for_ddl_low (trx=trx@entry=0x7f0fb0e023c0) at /mariadb/10.6/storage/innobase/trx/trx0trx.cc:2273
#6  0x00005571732a721c in dict_sys_t::create_or_check_sys_tables (this=0x557174abf100 <dict_sys>) at /mariadb/10.6/storage/innobase/dict/dict0crea.cc:1375
#7  0x000055717310b00b in srv_start (create_new_db=<optimized out>) at /mariadb/10.6/storage/innobase/srv/srv0start.cc:1909
#8  0x0000557172d1ce86 in innodb_init (p=<optimized out>) at /mariadb/10.6/storage/innobase/handler/ha_innodb.cc:4316
```
In MariaDB Server 10.6, this very first transaction with id 1 would write 27 undo log records (all of them `TRX_UNDO_INSERT_REC`, which does not trigger this code). The transaction would not invoke `btr_page_split_and_insert()` either.